### PR TITLE
Add pet clinic spring boot application

### PR DESCRIPTION
**Pull request by shalauniou**

### CODE REVIEW

File: src/main/java/com/epam/petclinic/clinic/controller/AnimalController.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import com.epam.petclinic.clinic.repository.AnimalRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RestController;
+
+/**
+ * Animal controller.
+ * <p/>
+ * Copyright (C) 2016 copyright.com
```
Comment (Andrei-Khadziukou) [inline src/main/java/com/epam/petclinic/clinic/controller/AnimalController.java]: Remove line

File: README.md [hunk 2]
```diff
@@ -1 +1,5 @@
 # petclinic-clinic
+Create data base with name petclinic
+Set postgres password in the application.properties for liquibase.password
```
Comment (Andrei-Khadziukou) [inline README.md]: we use *.yml

File: src/main/java/com/epam/petclinic/clinic/controller/AnimalController.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ * <p/>
+ * Date: 4/18/2017
+ *
+ * @author Stanislau Halauniou
+ */
+@RestController
+@RequestMapping("/api/animals")
```
Comment (Andrei-Khadziukou) [inline src/main/java/com/epam/petclinic/clinic/controller/AnimalController.java]: Do we need '/api'?

File: .gitignore [hunk 4]
```diff
@@ -0,0 +1,6 @@
+*.iml
```
Comment (Andrei-Khadziukou) [inline .gitignore]: remove from commit

File: src/main/java/com/epam/petclinic/clinic/controller/ClinicController.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+/**
+ * Clinic controller.
+ * <p>
+ * Date: 2017-04-13
+ *
+ * @author Stanislau_Halauniou
+ */
+@RestController
+@RequestMapping("/api/clinics")
```
Comment (Andrei-Khadziukou) [inline src/main/java/com/epam/petclinic/clinic/controller/ClinicController.java]: What do you think about specifying consume, produce parameters?

File: src/main/java/com/epam/petclinic/clinic/controller/OfferController.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ * Offer controller.
+ * <p/>
+ * Date: 4/12/2017
+ *
+ * @author Stanislau Halauniou
+ */
+//TODO: Might be removed
+@RestController
+@RequestMapping("/offer")
```
Comment (Andrei-Khadziukou) [inline src/main/java/com/epam/petclinic/clinic/controller/OfferController.java]: I think we should use plural form

File: src/main/java/com/epam/petclinic/clinic/controller/ClinicController.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    private OfferRepository offerRepository;
+
+    //TODO: add transaction, investigate cascading
+    @RequestMapping(value = "/", method = RequestMethod.POST)
+    public Clinic create(@RequestBody Clinic clinic) {
+        Clinic createdClinic = clinicRepository.save(clinic);
+        List<Offer> offers = clinic.getOffers();
+        offers.stream().forEach(offer -> offer.setClinic(clinic));
+        offerRepository.save(offers);
+        clinic.setId(createdClinic.getId());
```
Comment (Andrei-Khadziukou) [inline src/main/java/com/epam/petclinic/clinic/controller/ClinicController.java]: Why don't we just return createdClinic?

File: src/main/java/com/epam/petclinic/clinic/controller/ClinicController.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        clinicRepository.delete(id);
+    }
+
+    @RequestMapping(value = "/", method = RequestMethod.GET)
+    public Iterable<Clinic> getAll() {
+        return clinicRepository.findAll();
+    }
+
+    @RequestMapping(value = "/{id}", method = RequestMethod.GET)
+    public Clinic getClinicWithOffers(@PathVariable String id) {
```
Comment (Andrei-Khadziukou) [inline src/main/java/com/epam/petclinic/clinic/controller/ClinicController.java]: If clinic doesn't have offers it won't be displayed, will it?

File: src/main/java/com/epam/petclinic/clinic/controller/ClinicController.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        List<Offer> offers = clinic.getOffers();
+        offers.stream().forEach(offer -> offer.setClinic(clinic));
+        offerRepository.save(offers);
+        clinic.setId(createdClinic.getId());
+        return clinic;
+    }
+
+    //TODO: add transaction, investigate cascading
+    @RequestMapping(value = "/{id}", method = RequestMethod.PUT)
+    public Clinic update(@RequestBody Clinic clinic, @PathVariable String id) {
```
Comment (Andrei-Khadziukou) [inline src/main/java/com/epam/petclinic/clinic/controller/ClinicController.java]: Let's update only clinic fields (address, name and so on) and manage association between clinic and offer by 2 different calls (add/remove clinic offer)

File: src/main/java/com/epam/petclinic/clinic/controller/AnimalController.java (hunk 3 above)
Comment (shalauniou) [reply src/main/java/com/epam/petclinic/clinic/controller/AnimalController.java]: No matter, done according to the https://docs.google.com/document/d/1YFDeqRO4UQ7oDVkt96fEH3Tm2nPVcH8miBbgo9zxga8/edit#

File: src/main/java/com/epam/petclinic/clinic/controller/OfferController.java (hunk 6 above)
Comment (shalauniou) [reply src/main/java/com/epam/petclinic/clinic/controller/OfferController.java]: Should be removed at all

File: src/main/java/com/epam/petclinic/clinic/controller/ClinicController.java (hunk 9 above)
Comment (shalauniou) [reply src/main/java/com/epam/petclinic/clinic/controller/ClinicController.java]: According to the doc it should be one request

File: src/main/java/com/epam/petclinic/clinic/controller/AnimalController.java (hunk 3 above)
Comment (Andrei-Khadziukou) [reply src/main/java/com/epam/petclinic/clinic/controller/AnimalController.java]: See notes before table

File: src/main/java/com/epam/petclinic/clinic/controller/ClinicController.java (hunk 9 above)
Comment (Andrei-Khadziukou) [reply src/main/java/com/epam/petclinic/clinic/controller/ClinicController.java]: Could you please add TODO to think about my proposal and let's leave it as it is for now

File: src/main/java/com/epam/petclinic/clinic/controller/ClinicController.java (hunk 7 above)
Comment (shalauniou) [reply src/main/java/com/epam/petclinic/clinic/controller/ClinicController.java]: According to the doc, might be discussed later

File: src/main/java/com/epam/petclinic/clinic/controller/ClinicController.java (hunk 8 above)
Comment (shalauniou) [reply src/main/java/com/epam/petclinic/clinic/controller/ClinicController.java]: I don't know. Suggest to ask UI team

File: build.gradle [hunk 10]
```diff
@@ -27,18 +27,25 @@ dependencies {
...
 }
 
 dependencyManagement {
     imports {
         mavenBom "org.springframework.cloud:spring-cloud-dependencies:Dalston.RC1"
     }
 }
+
+test {
+    systemProperty "spring.profiles.active", "test"
```
Comment (Andrei-Khadziukou) [inline build.gradle:55]: Why do we need that?

File: src/main/java/com/epam/petclinic/clinic/controller/AnimalController.java (hunk 3 above)
Comment (Andrei-Khadziukou) [reply src/main/java/com/epam/petclinic/clinic/controller/AnimalController.java]: Any updates?

File: src/main/java/com/epam/petclinic/clinic/model/Animal.java [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+/**
+ * Animal entity.
+ * <p/>
+ * Date: 4/12/2017
+ *
+ * @author Stanislau Halauniou
+ */
+@Entity
+@Table(name="animal")
+@EqualsAndHashCode
```
Comment (Andrei-Khadziukou) [inline src/main/java/com/epam/petclinic/clinic/model/Animal.java]: How does it affect code?

File: src/main/java/com/epam/petclinic/clinic/model/Clinic.java [hunk 12]
```diff
@@ -0,0 +1,86 @@
+package com.epam.petclinic.clinic.model;
+
+/**
```
Comment (Andrei-Khadziukou) [inline src/main/java/com/epam/petclinic/clinic/model/Clinic.java]: Do we need that javadoc?

File: src/main/resources/liquibase/changelog.xml [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+        <rollback>
+        </rollback>
+    </changeSet>
+    <changeSet id="2017-04-18-17-00" author="Stanislau_Halauniou@epam.com">
+        <comment>Create unique constraint for offers</comment>
+        <addUniqueConstraint tableName="offer" columnNames="clinic_id, animal_id, clinic_service_id"
+                             constraintName="unique_offer"/>
+    </changeSet>
+
+</databaseChangeLog>
```
Comment (Andrei-Khadziukou) [inline src/main/resources/liquibase/changelog.xml]: Add new line, please

File: src/test/java/com/epam/petclinic/clinic/controller/ClinicControllerTest.java [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.junit.Assert.*;
+
+/**
+ * Tests {@link ClinicController}.
+ * <p/>
+ * Copyright (C) 2016 copyright.com
```
Comment (Andrei-Khadziukou) [inline src/test/java/com/epam/petclinic/clinic/controller/ClinicControllerTest.java]: Could you please clean up all javadoc-s?

File: build.gradle (hunk 10 above)
Comment (shalauniou) [reply build.gradle:55]: for chosing `spring.profiles.activeprofile` when run `test` task

File: src/main/java/com/epam/petclinic/clinic/model/Animal.java (hunk 11 above)
Comment (shalauniou) [reply src/main/java/com/epam/petclinic/clinic/model/Animal.java]: Replacement for equals and hashCode methods.  Should be discussed if we need it.

File: src/main/resources/application-test.yml [hunk 15]
```diff
@@ -0,0 +1,19 @@
+spring.profiles.active: test
```
Comment (Andrei-Khadziukou) [inline src/main/resources/application-test.yml]: Is it possible to move that class to test/resources

File: src/main/resources/liquibase/changelog.xml [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+        <addForeignKeyConstraint baseTableName="offer" baseColumnNames="animal_id"
+                                 constraintName="animal_id_constraint"
+                                 referencedTableName="animal" referencedColumnNames="id"/>
+        <addForeignKeyConstraint baseTableName="offer" baseColumnNames="clinic_id"
+                                 constraintName="service_id_constraint"
+                                 referencedTableName="clinic" referencedColumnNames="id"/>
+        <addForeignKeyConstraint baseTableName="offer" baseColumnNames="clinic_service_id"
+                                 constraintName="clinic_service_id_constraint"
+                                 referencedTableName="clinic_service" referencedColumnNames="id"/>
+        <rollback>
```
Comment (Andrei-Khadziukou) [inline src/main/resources/liquibase/changelog.xml]: Could you please check rollbacks are working

File: src/main/resources/application-test.yml (hunk 15 above)
Comment (shalauniou) [reply src/main/resources/application-test.yml]: Yes, moved

File: tasks.gradle [hunk 17]
```diff
@@ -0,0 +1,64 @@
+import groovy.sql.Sql
+
+String username = "${username}".toString()
```
Comment (Andrei-Khadziukou) [inline tasks.gradle]: I don't think we need toString()

File: tasks.gradle [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+URLClassLoader loader = GroovyObject.class.classLoader
+
+configurations.driver.each {File file ->
+    loader.addURL(file.toURL())
+}
+
+task createDatabase() {
+    doLast{
+        String createUserQuery = "create user ${username}".toString()
+        Sql.withInstance('jdbc:postgresql://localhost:5432/', admin,
```
Comment (Andrei-Khadziukou) [inline tasks.gradle]: could you make it configurable

File: tasks.gradle (hunk 17 above)
Comment (Andrei-Khadziukou) [reply tasks.gradle]: Do we really need that local variables?

File: src/main/java/com/epam/petclinic/clinic/model/Offer.java [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+    private String id;
+
+    private Clinic clinic;
+
+    private Animal animal;
+
+    private ClinicService clinicService;
+
+
+    //--------Getters---------
```
Comment (Andrei-Khadziukou) [inline src/main/java/com/epam/petclinic/clinic/model/Offer.java]: Do we need such comments?

File: README.md [hunk 20]
```diff
@@ -2,3 +2,10 @@
 
 ## Configuration
 * Add `management.security.enabled: false` property into `application.yml` file if you need to connect to secured endpoints from actuator (don't commit that configuration) 
+
+# Running
+Create database : configure gradle.properties
+run gradlew createDatabase(for dropping run: gradlew dropDatabase)
+Disable @EnableEurekaClient(move temporary bootstrap.yml) or run it
```
Comment (Andrei-Khadziukou) [inline README.md]: Provide a link to the app which should be ran

File: src/main/java/com/epam/petclinic/clinic/controller/AnimalController.java [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+import com.epam.petclinic.clinic.model.Animal;
+import com.epam.petclinic.clinic.repository.AnimalRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RestController;
+
+/**
+ * Animal controller.
+ * <p/>
```
Comment (Andrei-Khadziukou) [inline src/main/java/com/epam/petclinic/clinic/controller/AnimalController.java]: Follow single approach... Use `<p/>` or `<p>`

File: src/main/resources/liquibase/changelog.xml [hunk 22]
```diff
@@ -0,0 +1,149 @@
...
+                                 constraintName="service_id_constraint"
+                                 referencedTableName="clinic" referencedColumnNames="id"/>
+        <addForeignKeyConstraint baseTableName="offer" baseColumnNames="clinic_service_id"
+                                 constraintName="clinic_service_id_constraint"
+                                 referencedTableName="clinic_service" referencedColumnNames="id"/>
+    </changeSet>
+
+    <changeSet id="2017-04-13-15-00" author="Stanislau_Halauniou@epam.com">
+        <comment>Populate tables</comment>
+            <insert tableName="animal">
```
Comment (Andrei-Khadziukou) [inline src/main/resources/liquibase/changelog.xml]: correct formatting here

**Pull request merged**